File: frontend/src/types/operation.ts
export interface Operation {
  type: 'INSERT' | 'DELETE' | 'RETAIN';
  position: number;
  content?: string;
  length: number;
  userId: string;
  timestamp: number;
  revision: number;
  operationId?: string; // 클라이언트 내부 사용 (서버 전송 시 제외)
}

export interface DocumentState {
  content: string;
  revision: number;
}

export interface OperationMessage {
  type: string;
  pageId: number;
  staffId: string;
  userName: string;
  operation?: Operation;
  initialContent?: string;
  revision?: number;
  content?: string; // 마크다운 버튼용 전체 문서 내용
} 

File: src/main/java/com/wiki/dto/Operation.java
package com.wiki.dto;

import lombok.Data;
import lombok.AllArgsConstructor;
import lombok.NoArgsConstructor;

@Data
@AllArgsConstructor
@NoArgsConstructor
public class Operation {
    
    public enum Type {
        INSERT, DELETE, RETAIN
    }
    
    private Type type;
    private int position;
    private String content;
    private int length;
    private String userId;
    private long timestamp;
    private long revision;
    
    public static Operation insert(int position, String content, String userId, long revision) {
        Operation op = new Operation();
        op.type = Type.INSERT;
        op.position = position;
        op.content = content;
        op.length = content.length();
        op.userId = userId;
        op.timestamp = System.currentTimeMillis();
        op.revision = revision;
        return op;
    }
    
    public static Operation delete(int position, int length, String userId, long revision) {
        Operation op = new Operation();
        op.type = Type.DELETE;
        op.position = position;
        op.length = length;
        op.userId = userId;
        op.timestamp = System.currentTimeMillis();
        op.revision = revision;
        return op;
    }
} 